rand_os/src/solarish.rs

```diff
@@ -110,18 +112,50 @@ impl OsRngImpl for OsRng {
     }
 }
 
-fn getrandom(buf: &mut [u8], blocking: bool) -> libc::c_long {
-    extern "C" {
-        fn syscall(number: libc::c_long, ...) -> libc::c_long;
-    }
+#[cfg(target_os = "illumos")]
+type GetRandomFn = unsafe extern fn(*mut u8, libc::size_t, libc::c_uint)
+    -> libc::ssize_t;
+#[cfg(target_os = "solaris")]
+type GetRandomFn = unsafe extern fn(*mut u8, libc::size_t, libc::c_uint)
+    -> libc::c_int;
+
+// Use dlsym to determine if getrandom(2) is present in libc.  On Solarish
+// systems, the syscall interface is not stable and can change between
+// updates.  Even worse, issuing unsupported syscalls will cause the system
+// to generate a SIGSYS signal (usually terminating the program).
+// Instead the stable APIs are exposed via libc.  Cache the result of the
+// lookup for future calls.  This is loosely modeled after the
+// libstd::sys::unix::weak macro which unfortunately is not exported.
+fn fetch() -> Option<&'static GetRandomFn> {
+    static FPTR: AtomicUsize = AtomicUsize::new(1);
 
-    const SYS_GETRANDOM: libc::c_long = 143;
+    unsafe {
+       if FPTR.load(Ordering::SeqCst) == 1 {
+            let name = "getrandom\0";
+            let addr = libc::dlsym(libc::RTLD_DEFAULT,
+                                   name.as_ptr() as *const _) as usize;
+            FPTR.store(addr, Ordering::SeqCst);
+       }
+
+       if FPTR.load(Ordering::SeqCst) == 0 {
+           return None;
+       } else {
+           mem::transmute::<&AtomicUsize, Option<&GetRandomFn>>(&FPTR)
```

Comment:
Sorry, but this is horrible! The internal layout of `AtomicUsize` is undefined, i.e. changes to Rustc could break this, besides which you are not using the atomic operations here. And are you even transmuting the value in the FPTR or its address?

In general, prefer pointer casts over `transmute`; they are "less unsafe". Also, in this case you already check for the `None` case. Lets rearrange a little:
```rust
let ptr = FPTR.load(Ordering::SeqCst);
if ptr == 0 {
    None
} else {
    let ptr = ptr as *const GetRandomFn;
    Some(unsafe{ &*ptr })
}
```

Reply:
Unfortunately the suggested change segfaults -- since the GetRandomFn type is fn, it is already a pointer, so `*const GetRandomFn` I think ends up as a pointer to a pointer and the return value ends up being the address of temporary instead of the function.  I suspect if the symbol being looked up was a variable instead of a function, this would probably work.

I tried to find any examples of using libc::dlsym and casting the result, and could not -- all of them end up using transmute.  Asking around a bit, no one is aware of a way to cast the results of libc::dlsym for functions -- they always end up having to use transmute (even libstd ends up using transmute instead of casting for similar situations).

I also tried variations to see if I could get something to work with the cast (such as `let ptr = ptr as GetRandomFn`), which failed to compile because it was a non-trivial cast (which I suspect goes back to why all the examples end up using transmute).

Reply:
Okay then, I was wrong on this; my apology for the diversion and confusion over function pointers.